psi4/src/psi4/libfock/DirectDFJK.cc

```diff
@@ -0,0 +1,1785 @@
+     * the number of blocks over which AO ERI's need to be calculated */
+    charges_f = charges_f / 2.0;
+
+    charges_z = static_cast<size_t>(charges_f);
+    ++charges_z;
+
+    // D
+    our_needs += nbf_ * nbf_;
+    free_memory_ -= nbf_ * nbf_;
+
+    // C_a
+    our_needs += nbf_ * charges_z;
+    free_memory_ -= nbf_ * charges_z;
+    // C_b
+    our_needs += nbf_ * charges_z;
+    free_memory_ -= nbf_ * charges_z;
+    // K
+    our_needs += nbf_ * nbf_;
+    free_memory_ -= nbf_ * nbf_;
+    // J
+    our_needs += nbf_ * nbf_;
+    free_memory_ -= nbf_ * nbf_;
+    //[ J^{-1} ]
+    our_needs += naux_ * naux_;
+    free_memory_ -= naux_ * naux_;
+    //[ J^{-\frac{1}{2} ]
+    our_needs += naux_ * naux_;
+    free_memory_ -= naux_ * naux_;
+    blocks = (2 * (nbf_ * nbf_ * naux_)) / free_memory_ + 1;
+    ABX_block_size_ = ((naux_ / blocks) + 1) * nbf_ * nbf_;
+
+    total_needs_ = our_needs;  // + 2*ABX_block_size_;
+}
+size_t DirectDFJK::memory_estimate() {
+    if (Qpq_) {
+        prepare_Q_blocks();
+    }
+
+    our_needs();
+    return total_needs_;
+}
+
+void DirectDFJK::prepare_metric_power(double power) {
+    SharedMatrix my_met;
+    auto coul_met = std::make_shared<FittingMetric>(auxiliary_, true);
+    coul_met->form_fitting_metric();
+    my_met = coul_met->get_metric();
+    if (!(1e-13 > fabs(1.0 - power))) {
+        my_met->power(power, condition_);
+    }
+
+    met_cols_.push_back(my_met->ncol());
+    met_rows_.push_back(my_met->nrow());
+    metric_.push_back(my_met);
+    met_powers_.push_back(power);
+
+    while (power < 0.0) {
+        power = power + 1.0;
+    }
+}
+
+double* DirectDFJK::get_metric_power(double power) {
+    bool pow_on = false;
+    double ret_power;
+    size_t met_ind;
+    for (size_t met_iter = 0; met_iter < metric_.size(); met_iter++) {
+        if (std::fabs(power - met_powers_[met_iter]) < condition_) {
+            met_ind = met_iter;
+            pow_on = true;
+        }
+    }
+    if (!pow_on) {
+        met_ind = metric_.size();
+        SharedMatrix my_met;
+        auto coul_met = std::make_shared<FittingMetric>(auxiliary_, true);
+        coul_met->form_fitting_metric();
+        my_met = coul_met->get_metric();
+        my_met->power(power, condition_);
+        metric_.push_back(my_met);
+        met_powers_.push_back(power);
+    }
+    return metric_[met_ind]->pointer()[0];
+}
+
+// Computes the LU decomposition of the coulomb metric and stores it in
+//   member vector CMPQ_LU_.
+//   pivots are storec in vector PERMUTE_
+void DirectDFJK::get_met() {
+    auto coul_met = std::make_shared<FittingMetric>(auxiliary_, true);
+    int info;
+
+    coul_met->form_fitting_metric();
+    SharedMatrix my_met;
+    my_met = coul_met->get_metric();
+    PERMUTE_.resize(naux_);
+    CMPQ_LU_.resize(naux_ * naux_);
+
+    CMPQ_inv_.resize(naux_ * naux_);
+
+//    double ZERO[8] = {0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0};
```

Comment:
sorry about that.

Reply:
Also I can't reply to your previous comment:

"Instead of LU you could try the FittingMetric::form_cholesky_inverse() routine. That is closer to what is done by Ahlrichs and also used e.g. in ORCA's RIJK implementation. With available optimized RI basis sets that works well and is fast.

Heuristically auto-generated (over-complete) RI basis sets as available in ORCA can lead to numerical problems and pivoting is needed"

so I will here. I did try the cholesky inverse when I was first writing the code, and conditioning errors persisted. That's why I went with the LU decomposition with partial pivoting.